### DISCUSSION TITLE
Langium 3 and ERR_UNHANDLED_REJECTION

### DISCUSSION THREAD

**Discussion by cdietrich** [eclipse-langium/langium#1342] [Q&A] [answered]
i see this when starting our vscode langium based extension updated to langium 3x.
```
node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^

[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "Symbol(OperationCancelled)".] {
  code: 'ERR_UNHANDLED_REJECTION'
}
Thrown at:
    at getErrorWithoutStack (node:internal/process/promises:320:15)
    at generateUnhandledRejectionError (node:internal/process/promises:340:15)
    at processPromiseRejections (node:internal/process/promises:286:24)
    at processTicksAndRejections (node:internal/process/task_queues:96:32)
```
    
do you have any hints on how to analyze?

i can see interruptAndCheck does throw the symbol. but where should this be catched and handled?

**Comment by msujew** [comment]
@cdietrich Everything that is able to throw this error runs inside of the document builder, which in itself runs inside of the [`WorkspaceLock`](https://github.com/eclipse-langium/langium/blob/main/packages/langium/src/workspace/workspace-lock.ts). The workspace lock however, handles all errors that are **caught inside of it** without ever throwing itself.

It might be an error similar to [here](https://github.com/eclipse-langium/langium/discussions/1112) where a missing `await` prevented the workspace lock from catching the error. Are you sure that you always await any potential throwing operation?

**Reply by cdietrich** [reply]
hmmm i see it in initializeWorkspace. but i need to diff further. somehow eslint shows nothing ...

**Reply by cdietrich** [reply]
is there anything specific that has changed in this area in 3x?

**Reply by msujew** [reply]
We rewrote the workspace lock, see https://github.com/eclipse-langium/langium/pull/1310. Aside from that, nothing should have changed. We have tested the currently available next version already in some of our projects, but noticed no issues.

**Reply by msujew** [reply]
Note that you might want to instruct your debugger to catch uncaught exceptions, maybe that could help track down the source of the error.

**Reply by cdietrich** [reply]
the workspace we test with is quite big. in a smaller i dont see an issue.
i wonder if there is something special about initializeWorkspace

**Reply by cdietrich** [reply]
the point is this is the only one it catches what i dont understand. 
where should initializeWorkspace.interruptAndCheck be handled? where can i set this breakpoint.

**Reply by cdietrich** [reply]
<img width="1184" alt="grafik" src="https://github.com/eclipse-langium/langium/assets/238736/ea81872a-c404-41d4-acbc-bdfa4498c59a">
this looks bad, correct?

but of course it almost never stops there ....

**Reply by cdietrich** [reply]
manually patched the bundled js. and i see the same.
for Symbol(OperationCancelled) it goes to the else.
i wonder if bundlers have problems with duplicate symbols. we use esbuild

**Reply by cdietrich** [reply]
indeed seems to be a bunding problem. we need to check ...

**Reply by msujew** [reply]
> i wonder if bundlers have problems with duplicate symbols. we use esbuild

Yes, symbols are unique pointers. If you somehow manage to duplicate a symbol, performing a `===` will fail against the duplicated one. We use esbuild ourselves though and haven't encountered this issue yet.

**Reply by msujew** [reply]
@cdietrich I assume that your package manager somehow duplicated the `langium` package and it's now getting bundled twice into your app. See https://github.com/evanw/esbuild/issues/475#issuecomment-1480312439. Esbuild should tell you where these duplicated packages come from in the bundled code.

**Reply by cdietrich** [reply]
we have one project where we have the language and bundle the module (with langium)
and then a second project that uses the bundle from 1 and also uses langium.
naively this should work. but i need to check.
maybe we also need to build 2 different bundles in 1

**Reply by cdietrich** [reply, marked as answer]
fixing our esm bundle not to include langium solved the problem. we will bundle it in csm only now.
